ssl/statem/extensions_clnt.c

```diff
@@ -679,12 +679,85 @@ EXT_RETURN tls_construct_ctos_early_data(SSL *s, WPACKET *pkt,
                                          unsigned int context, X509 *x,
                                          size_t chainidx, int *al)
 {
+    const unsigned char *id;
+    size_t idlen = 0;
+    SSL_SESSION *psksess = NULL;
+    SSL_SESSION *edsess = NULL;
+    const EVP_MD *handmd = NULL;
+
+    if (s->hello_retry_request)
+        handmd = ssl_handshake_md(s);
+
+    if (s->psk_use_session_cb != NULL
+            && (!s->psk_use_session_cb(s, handmd, &id, &idlen, &psksess)
+                || (psksess != NULL
+                    && psksess->ssl_version != TLS1_3_VERSION))) {
+        SSL_SESSION_free(psksess);
+        SSLerr(SSL_F_TLS_CONSTRUCT_CTOS_EARLY_DATA, SSL_R_BAD_PSK);
+        return EXT_RETURN_FAIL;
+    }
+
+    SSL_SESSION_free(s->psksession);
+    s->psksession = psksess;
+    if (psksess != NULL) {
+        OPENSSL_free(s->psksession_id);
+        s->psksession_id = OPENSSL_memdup(id, idlen);
+        if (s->psksession_id == NULL) {
+            SSLerr(SSL_F_TLS_CONSTRUCT_CTOS_EARLY_DATA, ERR_R_INTERNAL_ERROR);
+            return EXT_RETURN_FAIL;
+        }
+        s->psksession_id_len = idlen;
+    }
+
     if (s->early_data_state != SSL_EARLY_DATA_CONNECTING
-            || s->session->ext.max_early_data == 0) {
+            || (s->session->ext.max_early_data == 0
+                && (psksess == NULL || psksess->ext.max_early_data == 0))) {
         s->max_early_data = 0;
         return EXT_RETURN_NOT_SENT;
     }
-    s->max_early_data = s->session->ext.max_early_data;
+    edsess = s->session->ext.max_early_data != 0 ? s->session : psksess;
+    s->max_early_data = edsess->ext.max_early_data;
+
+    if ((s->ext.hostname == NULL && edsess->ext.hostname != NULL)
+            || (s->ext.hostname != NULL
+                && (edsess->ext.hostname == NULL
+                    || strcmp(s->ext.hostname, edsess->ext.hostname) != 0))) {
+        SSLerr(SSL_F_TLS_CONSTRUCT_CTOS_EARLY_DATA,
+               SSL_R_INCONSISTENT_EARLY_DATA_SNI);
+        return EXT_RETURN_FAIL;
+    }
+
+    if ((s->ext.alpn == NULL && edsess->ext.alpn_selected != NULL)) {
+        SSLerr(SSL_F_TLS_CONSTRUCT_CTOS_EARLY_DATA,
+               SSL_R_INCONSISTENT_EARLY_DATA_ALPN);
+        return EXT_RETURN_FAIL;
```

Comment:
Why would it not be legal? The server may not accept our PSK and revert to a full handshake. In that case we may wish it to look at the ALPN list, even though we didn't use it for the early_data (which the server ignored).

Reply:
Well, I did not spend too much time thinking about this question, so may well be wrong.
An assumption I had for my thinking was that we could be pretty sure the server supports ALPN negotiation to some extent (but I did not check whether that is actually required by the spec); similarly I was assuming that a server that supports ALPN and is presented with a list of protocols should pick one of them (and in fact, RFC 7301 seems to say that the server MUST send a fatal alert if ALPN negotiation fails).  So, if we've talked to this server before, and this server supports ALPN, we'd have to have been doing something extremely weird/suspect in order for the server's resumption state to not have an associated ALPN value.
But I consider this discussion to be a side note with respect to the substance of the pull request, and it can be resolved (or ignored) separately.

Reply:
We are talking abut the scenario where the server previously issues a PSK ticket with no associated ALPN (which is what we are using for sending early_data). It does not seem unreasonable to me that, coming back some time later, we find that the server does now support ALPN. The configuration may have been updated, the server upgraded, or we may have been talking to a cluster of servers with varying capabilities.

Reply:
Okay, I'll drop it.